pandas/_libs/index.pyx

```diff
@@ -648,10 +648,13 @@ cdef class BaseMultiIndexCodesEngine:
         self.levels = levels
         self.offsets = offsets
 
-        # Transform labels in a single array, and add 1 so that we are working
-        # with positive integers (-1 for NaN becomes 0):
-        codes = (np.array(labels, dtype='int64').T + 1).astype('uint64',
+        # Transform labels in a single array, and add 2 so that we are working
+        # with positive integers (-1 for NaN becomes 1). This enables us to
+        # differentiate between values that are missing in other and matching
+        # NaNs. We will set values that are not found to 0 later:
+        codes = (np.array(labels, dtype='int64').T + 2).astype('uint64',
```

Comment:
For other engines, are we offsetting by 1?

It's a nice solutions, but I am a little worried about other having to remember a (maybe) 1 vs 2 shift. Is there a way to define a `null_shift` variable that can be used consistently

Reply:
This is is only necessary for the MultiIndex engine, since the NaNs are not in the levels. Other engines are matching NaNs directly.

Can set it via a variable if you like, would at least cover the size case and this here

Reply:
Gotcha. Yeah if there's any way to centralize & minimize the shifting of codes (via a variable or function), that would be great to make it easier to connect the `+1`s and `+2`s in different places

Reply:
moved the +2 into a variable, the others aren't practical I think